modules/queue/unique_queue_channel.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2020 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package queue
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"code.gitea.io/gitea/modules/log"
+)
+
+// ChannelUniqueQueueType is the type for channel queue
+const ChannelUniqueQueueType Type = "unique-channel"
+
+// ChannelUniqueQueueConfiguration is the configuration for a ChannelUniqueQueue
+type ChannelUniqueQueueConfiguration ChannelQueueConfiguration
+
+// ChannelUniqueQueue implements UniqueQueue
+//
+// It is basically a thin wrapper around a WorkerPool but keeps a store of
+// what has been pushed within a table
+type ChannelUniqueQueue struct {
+	*WorkerPool
+	lock     sync.Mutex
+	table    map[Data]bool
+	exemplar interface{}
+	workers  int
+	name     string
+}
+
+// NewChannelUniqueQueue create a memory channel queue
+func NewChannelUniqueQueue(handle HandlerFunc, cfg, exemplar interface{}) (Queue, error) {
+	configInterface, err := toConfig(ChannelUniqueQueueConfiguration{}, cfg)
+	if err != nil {
+		return nil, err
+	}
+	config := configInterface.(ChannelUniqueQueueConfiguration)
+	if config.BatchLength == 0 {
+		config.BatchLength = 1
+	}
+	queue := &ChannelUniqueQueue{
+		table:    map[Data]bool{},
+		exemplar: exemplar,
+		workers:  config.Workers,
+		name:     config.Name,
+	}
+	queue.WorkerPool = NewWorkerPool(func(data ...Data) {
+		for _, datum := range data {
+			queue.lock.Lock()
+			delete(queue.table, datum)
```

Comment:
The datum is removed from the table before it's handled. `checkAndUpdateStatus()` call to `.Has()` becomes pointless.

Reply:
Are you taking about in check.go?

Yes, this replicates current behaviour. Once you start handling a queue datum you have to assume that any further additions may be because of updates to that datum - hence it should be removed from the unique queue table at that point.

Reply:
Yeah, but now you're not preventing two checks from running at the same time! 🤔

Reply:
Hmm... need to think about this one. The logic in the check PR function is pretty bad in any case and needs a refactor. Part of me wonders if we should fix it in another PR.

Reply:
Right, as I've written before the logic here is broken. We really need your temporary lock thing to do the status changes but I think (!) if we change the following then it should improve things:

https://github.com/go-gitea/gitea/blob/5ab601b1bd8b0e0d71f1d3b892dc21dd448be791/services/pull/check.go#L30-L42

This needs to only update to StatusChecking if status is not StatusMerged.

Then... 

https://github.com/go-gitea/gitea/blob/5ab601b1bd8b0e0d71f1d3b892dc21dd448be791/services/pull/check.go#L199-L220

1. We should only get the PR if the status is not merged.
2. Then we can drop the check in 208
3. and in setMerged we need to only update if not already set merged.